### DISCUSSION TITLE
Potential Feature: snapcast

### DISCUSSION THREAD

**Discussion by bs42** [micro-nova/AmpliPi#75] [Ideas] [answered]
[Snapcast](https://github.com/badaix/snapcast) support

> Snapcast is a multi-room client-server audio player, where all clients are time synchronized with the server to play perfectly synced audio.

The idea is that the AmpliPi could be either a client or a server for snapcast.  For larger locations you could have 2 AmpliPis in different locations powering lots of zones but minimizing wiring issues/distance.  For some homes you could have the home on an AmpliPi+extenders, and portable speaker(s) or a pi+amp powering a backyard zone.

**Comment by linknum23** [comment, marked as answer]
Snapcast is an awesome open source project for distributed wireless speakers! We've been following their progress for a while now, it looks really useful for apartment and other setups where you would use mostly wireless speakers.

In the future AmpliPi could host a snapcast server to add support for wireless speakers. The only issue with this, using the current hardware, is that only digital sources and streams would be playable (the analog RCA inputs connect directly to the audio muxes). We think this would still add a lot of flexibility for mixed speaker setups.

Using AmpliPi as a snapcast client is a cool idea as well. The AmpliPi could potentially host up to 4 snapcast clients. In this case we would recommend routing each snapcast client to a group of zones, since each of AmpliPi's zones wouldn't be individually discoverable without significant changes to how snapcast's protocol works.

Also although synchronizing 2 AmpliPi's over the network might be possible, we have not designed the software with this use case in mind. 

Which of these would be the most useful? Feel free to make an issue for either client or server support.

**Reply by linknum23** [reply]
I moved this to discussions, since it seemed like a better fit.

**Reply by bs42** [reply]
Based on what you've said about the hardware RCA inputs, I think that using the AmpliPi as a server for snapcast becomes potentially confusing as there are now limits on which sources can be cast, etc.

I would think that the simplest would be to add snapcast as a new digital input only.  It minimizes any changes to the AmpliPi as it's just a new digital source.  Then it's up to the user to setup their own server and their snapcast enabled zones can use that server.

An advanced user who might be considering 2 AmpliPis could setup their own snapcast server and have that one digital stream sent to whichever speaker groups on the 2 AmpliPis they wanted.

My use case is potentially a raspberry pi + amp for my backyard zone and the AmpliPi for the ceiling speakers in the house.  It would still be simple to do so with a snapcast server with mopidy on my home server instead of the AmpliPi.

**Reply by linknum23** [reply]
I have a couple of ideas to make your configuration a little easier:
1. Obviously wiring outdoor speakers would make the configuration a lot simpler (assuming the wiring itself wasn't that complicated)
2. Have you considered using the new preamp outputs to a wireless transmitter? This could be advanced as a purpose-built rf transmitter + speaker pair, or be a wifi-based solution with a raspberry pi running snapcast server with a ADC (connected to the preamp)  --->  a raspberry pi with a DAC + AMP connected to the outdoor speakers.

**Reply by bs42** [reply]
1. It's a possibility, the wire length gets to be pretty long, I have a big back yard, but it's on my list to consider.
2. I had not considered a traditional wireless transmitter.  The pi running a snapcast server and a pi running snapcast client was my scenario, I just added in the amplipi as a snapcast client so the amplipi could get the same source rather than using one of the RCA connectors.

**Reply by linknum23** [reply]
Another option you could consider is using an audio balun for transmission. This could be used to convert the preamp outputs to a signal that can be transmitted over longer distances using cat5. This would need to be connected to outdoor powered speakers. Here's one on Amazon: https://www.amazon.com/dp/B08JQHT14.
One of our engineers, @mabowers, is planning on using this to connect his TV's output into the AmpliPi in his basement rack.
